### ISSUE TITLE
what should bounding_modes do in implementations and in i-score

### ISSUE THREAD

**Issue by bltzr** [OSSIA/libossia#105]
For now, e.g. for an ossia.parameter \@type float \@bounding_mode wrap in pd (or Max): 
-- when manually set from pd (with a number box at the input)  outside of the bound, the value goes out unchanged of the object outlet: e.g. in=12 -> out=12 / in i-score, the value shows up as 2
-- when manually set from i-score (from the device explorer), the value is processed through the bounding_mode: when set as 12 in i-score, ossia.param outputs 2

What I think would make more sense (and also allow to really get the benefits from the bounding_modes would be that, instead):
-- when manually set from pd (with a number box at the input) outside of the bound, the value is processed at the output of the object: e.g. in=12 -> out=2, but in i-score, the value still shows up as 12
-- when manually set from i-score (from the device explorer), the value is processed through the bounding_mode: when set as 12 in i-score, ossia.param outputs 2 (same as currently)
This would allow  e.g. in the case of polar coordinates, to record automations for multiples rotations, e.g. set angle at 720 in pd, snapshot in i-score, set angle at -360 in pd, snapshot and create an automation in i-score -> the automation will make 3 full rotations

In other words, that would mean that the bounding_mode processing would happen *just before* ossia.param's output, and not before - in other words the registered value of the parameter is not touched by the bounding_mode.
AFAIR this is not how it happened in Jamoma, excepted when sending a ramp, e.g. (/angle 720 ramp 2000) - which is somehow what we want to be able to reproduce here with i-score... but which then requires a different strategy...

Would that make sense ?
opinions @lossius @jln- @jcelerier @avilleret ???

**Comment by jcelerier** [comment]
I think that it could be done for instance in ø.parameter by doing : 

```C++
  local_address->add_callback([=](const ossia::value& v) { 
      setValue(
        net::filter_value(
          local_address->get_domain(), 
          v,
          local_address->get_bounding()));
  });
```

in ossia-pd/src/parameter.cpp:226

**Comment by bltzr** [comment]
@jln- what do you think about this ?

**Comment by avilleret** [comment]
this makes sense for me
and moreover, if the value set in Pd doesn't go through the filter this is a bug that could be dangerous
for example when setting volume in dB if it is not clipped you can add 100dB which could break your ears, your speakers, burn your computer and trig a nuclear war...
so I will fix this asap

**Comment by jcelerier** [comment]
:+1: for fixing nuclear wars

**Comment by jln-** [comment]
Sorry, I bookmarked this one but forgot to reply.

On the Max/Pd side, I agree that value output from ossia.param should comply to bounding_mode.

I really doubt about the i-score Device Explorer part, though. As a view of the namespace, the DE should imho display the value output by the parameters. Not doing so would mean that this view does not reflects the state the system currently is at. This doesn't sound very desirable to me. Thinking out loud, I'd say that entering 12 in the DE, automatically get clipped (or processed with actual bound mode) and changed to 2 as in the example proposed by Pascal.

However making values displayed in the Device Explorer stick to their actual value should not prevent from writing processes (ie automation) where a value sent from i-score exceed default range (thus allowing the behavior Pascal gave as an example). But this more sounds to me like a useful and handy hack rather than something we want to generalize.

Does this make sense ?

**Comment by bltzr** [comment]
I see your point, and it probably makes sense. thanks!

Since, we're thinking out loud here, wouldn't it be useful that there be a kind of "double display" of the values in i-score's DE, that would be also useful for unit conversion
i.e. if I have a parameter which internally deals with angles in radians, I could be able to manipulate it in degrees in the DE. Then, when e.g. setting the angle as 720 degrees, I would see that the result is actually 2π.
That probably isn't for tomorrow morning, but maybe that's something to include in our reflexions for future DE's design evolutions, since we're working on that ?

**Comment by avilleret** [comment]
could we close this @bltzr ?

**Comment by bltzr** [comment]
but, was what @jcelerier proposed above implemented ?
```
 local_address->add_callback([=](const ossia::value& v){ ... }
```
 I couldn't find it in the code...

**Comment by avilleret** [comment]
here : https://github.com/OSSIA/libossia/blob/master/OSSIA/ossia-pd/src/object_base.cpp#L216-L219

**Comment by bltzr** [comment]
OK, great, and that's implemented in Max too, I guess ?

**Comment by bltzr** [comment]
yup: https://github.com/OSSIA/libossia/blob/496777e8e3ee2dc49061c50de2dff62b826bb116/OSSIA/ossia-max/src/object_base.cpp#L221
